consensus/src/liveness/leader_reputation.rs

```diff
@@ -111,15 +107,18 @@ impl ActiveInactiveHeuristic {
 
 impl ReputationHeuristic for ActiveInactiveHeuristic {
     fn get_weights(&self, candidates: &[Author], history: &[NewBlockEvent]) -> Vec<u64> {
-        let set = history.iter().fold(HashSet::new(), |mut set, meta| {
-            set.insert(meta.proposer());
-            set.extend(meta.votes().into_iter());
-            set
-        });
-        candidates
-            .iter()
-            .map(|author| {
-                if set.contains(&author) {
+        let mut mask = vec![false; candidates.len()];
+        // set mask if it's in any of the event and ignore those with different length
+        for event in history {
+            if event.participants().len() == mask.len() {
```

Comment:
Will this do the right thing if `history` contains validator sets of the same size, but with different addresses? For example, if

- `candidates` is `[0xB, 0xC]`
- `history` is `[NewBlockEvent { votes: [0xA, 0xC] }, NewBlockEvent { votes: [0xA, 0xD] }]`

, it seems like `0xB` will get `active_weight` even though they have not voted on anything in `history`.

Reply:
great question! it's possible but unlikely since it needs at least two epochs to achieve same set with one different addresses. given it's just heuristic I don't think it matters much, if 0XB is not active its weight will drop soon.

Reply:
Makes sense. I think we could get the exact answer by comparing `candidates` and `event.participants`? But might be overkill if it's unlikely in practice + a heuristic anyway.

Reply:
it's also unlikely because we filtered the history by round, and round is reset every epoch. so most times it'd be 
`NewBlockEvent { round: 123456789, votes: [0xA, 0xC] } NewBlockEvent { round: 1, votes: [0xA, 0xD]}` and the first event will be filtered out.

Reply:
`event.participants` is just a vector of bool without actual addresses, in order to get the exact answer we'd have to pass the epoch number in NewBlockEvent so that we can filter by that but it seems overkill to me

Reply:
I agree these are edge cases, but that behavior Sam pointed out is odd (and basically random), whereas before we relied on correct historical data in all cases.

@zekun000 in any case, we'll have to update https://github.com/libra/libra/tree/master/specifications/consensus#leaderreputation with your point about the round reset by epochs have an impact on the filtering.

Reply:
I don't think it changes anything in the spec, it's all implementation details. The spec doesn't talk about how to represent a committed block at all.